src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,363 @@
+            {BinaryOperatorKind.Or, BinaryOperatorKind.And},
+            {BinaryOperatorKind.And, BinaryOperatorKind.Or},
+            {BinaryOperatorKind.ConditionalOr, BinaryOperatorKind.ConditionalAnd},
+            {BinaryOperatorKind.ConditionalAnd, BinaryOperatorKind.ConditionalOr},
+        };
+
+        protected abstract SyntaxNode GetIfStatement(TextSpan textSpan, SyntaxToken token, CancellationToken cancellationToken);
+        protected abstract SyntaxNode GetRootWithInvertIfStatement(Document document, SemanticModel model, SyntaxNode ifStatement, CancellationToken cancellationToken);
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract string GetTitle();
+        protected abstract bool IsFalseLiteralExpression(SyntaxNode expression);
+        protected abstract bool IsTrueLiteralExpression(SyntaxNode expression);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(textSpan.Start);
+
+            var ifStatement = GetIfStatement(textSpan, token, cancellationToken);
+            if (ifStatement == null)
+            {
+                return;
+            }
+
+            if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    GetTitle(),
+                    c => InvertIfAsync(document, ifStatement, c)));
+        }
+
+        private async Task<Document> InvertIfAsync(
+            Document document, 
+            SyntaxNode ifStatement, 
+            CancellationToken cancellationToken)
+        {
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            return document.WithSyntaxRoot(
+                // this returns the root because VB requires changing context around if statement
+                GetRootWithInvertIfStatement(
+                    document, model, ifStatement, cancellationToken));
+        }
+
+        internal SyntaxNode Negate(
+            SyntaxNode expression,
+            SyntaxGenerator generator,
+            ISyntaxFactsService syntaxFacts,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            if (syntaxFacts.IsParenthesizedExpression(expression))
+            {
+                return syntaxFacts.Parenthesize(
+                    Negate(
+                        syntaxFacts.GetExpressionOfParenthesizedExpression(expression),
```

Comment:
this expression likely need to have trivia stripped.

Reply:
Can you elaborate why the trivia should be stripped from the expression?  I added new tests with trivia around parentheses and other than this new issue, https://github.com/dotnet/roslyn/issues/26427, didn't have problems.

Reply:
ah, sorry.  i misread this.  i thought this was just negating the expression, and then adding the expression's trivia to the outside negation.  i see now that it's grabbing the expression *inside* the parens.  So this looks fine to me.